### DISCUSSION TITLE
Searching front-matter of all docs

### DISCUSSION THREAD

**Discussion by bradumbaugh** [fralau/mkdocs-macros-plugin#64] [Q&A] [answered]
Can a macro inspect the front-matter of all the Markdown files in my project's docs folder? I'm trying to write a `link_to` macro that receives a key, finds a doc that has a matching key in its front-matter, and returns an absolute link to that file. 

This is a great library, thanks! Enjoying playing around with it.

**Comment by fralau** [comment, marked as answer]
Thanks. 👍  

Yes you probably could write such a macro!  Here are a few pointers.

You could write a `main.py` program (in the top directory of your project), with a function that scan sthe yaml files in the [docs directory ](https://mkdocs-macros-plugin.readthedocs.io/en/latest/tips/#how-can-i-get-the-projects-document-directory)(`env.conf['docs_dir']` should contain its (relative) name; unless you are sure it's called `docs`)

It could then read all YAML headers .

```python
define_env(env):
    """
    This is the hook for defining variables, macros and filters

    - variables: the dictionary that contains the environment variables
    - macro: a decorator function, to declare a macro.
    """

    @env.macro
    def link_to(key:str) -> str:
          """Finds a doc that has a matching key in its front-matter, and returns an absolute link to that file"""
          ...
          return link
```

Here are the basic [instructions for writing macros](https://mkdocs-macros-plugin.readthedocs.io/en/latest/macros/#the-define_env-function).

To inspect the YAML front matters, you can check some good Python libraries, such as [python-frontmatter](https://github.com/eyeseast/python-frontmatter).

Let me know if that's sufficient to get you started?

**Reply by bradumbaugh** [reply]
Thanks @fralau, that's helpful! I'd thought that maybe the front-matter was already parsed and available, tucked away in some variable—but it makes sense that I could do that manually. That'd be an expensive task, though... if I built a dictionary mapping keys to links using the approach you describe, could I somehow keep it around  between macro invocations, so that it only needs to run once?

**Reply by fralau** [reply]
> Thanks @fralau, that's helpful! I'd thought that maybe the front-matter was already parsed and available, tucked away in some variable

🤔 @bradumbaugh it depends on what you want to do... [Each key/pair in the front matter is available out of the box within each markdown page, as a variable](https://mkdocs-macros-plugin.readthedocs.io/en/latest/pages/#local-page-level-variables-and-macros). 

```yaml
---
title: My special title
bottles:
  whine: 500
  beer: 123
---
```

You could write, within the page:

```
If have {{ bottles.whine }} bottles of whine.
```

You would have to describe exactly what result you want to achieve.

**Reply by bradumbaugh** [reply]
Thanks for bearing with me, @fralau, and for the helpful responses. I appreciate the examples. 

Basically, I'm trying to write a macro something like this: `link_to("My Link Text", "key-for-doc-a")`. That macro would return a Markdown link to Doc A (something like `[My Link Text](/path/to/doc/a)`. Doc A would have front-matter setting `key` to `key-for-doc-a`. Each of my docs would have a unique key.

A couple of questions:

1. Within a single macro call, can I directly access the front-matter for _all_ of my docs (without reading the files in and parsing them myself)? Or only for the page where the macro was called (I believe this is the case)?

2. There are at least a couple of ways I could build a dictionary mapping these per-doc keys to their corresponding URLs:

    - Manually read in all the docs and grab their front-matter
    - Inspect data already available in the `env` object (if it's there for all docs, see the first question)
 
Either way, this will be expensive, and I won't want to do it every time the `link_to` macro is called. So I'd like to construct that mapping only once—the first time the macro is called—and memoize it, saving it on the `env` object or somewhere else. This way, it'd be available for future macro calls (made from other docs in the `docs` folder) without the expense of rebuilding it. Is there a way to keep an object in existence across macro calls?

 Thanks! I'll pipe down here, don't mean to belabor this thread. Appreciate your interest, though.

**Reply by fralau** [reply]
> Thanks! I'll pipe down here, don't mean to belabor this thread. Appreciate your interest, though.

Please don't worry about that. I am interested in what you are doing, because it gives me a useful feedback.

If we take one step back: I would really like to understand you use case, and the flow of information?

Why do you want to make the information in the front page of _one page_, visible to _all_ other pages? Normally, the information in the front matter of a page, is for scope of that page only ("local"). That's why the plugin makes that information available only to the page.

If you want to store key-value pairs that are common to _all_ pages, there is a standard tool: the [configuration variables](https://mkdocs-macros-plugin.readthedocs.io/en/latest/pages/#configuration-variables), located in the `mkdocs.yml' file, under `extra`. They are also available out of the box.

Perhaps there is a good reason for you to want to "scope globally" the key/value pairs in the front-matters of pages  (i.e. to make them available to all other pages) 🤔   But I would need to understand it?

**Reply by bradumbaugh** [reply]
Thank you @fralau. Apologies for the slow response. I think you're right, and I probably want to use the extras. 

I'm trying to give each document a separate `key` (metadata), and to create a `link_to` macro that—given a `key` and some text—returns a Markdown link to the document associated with `key`. This would create a layer of abstraction for inter-document links, and keep them from breaking when filenames change.

I'd been thinking that maybe I could keep each document's `key` in its front-matter (rather than in the extras). That way, my macro could assemble at runtime a dictionary of key -> URL mappings, and `link_to` could use that dynamically generated list to create Markdown links.  This would prevent me from having to manually keep the `extras` up-to-date as I add / rename / delete docs. 

But it's probably not a huge problem to do that, and maybe I can just generate the extras with a script at build time.

Thank you again for your interest.

**Reply by fralau** [reply]
@bradumbaugh Thanks again. Your answers are helfpful, as they give me new insights. If I understand well, you want to implement a sort of **interwiki** system, where the definitions of the key/value pairs (prefixes) are specific to each page?

> **Interwiki links** are links to pages of other projects, using a prefixed internal link style. 
> A link like `[[Wikipedia:Main Page]]` will appear like this: `Wikipedia:Main Page` and redirect you to Wikipedia's Main Page. 
> ([Mediawiki's documentation](https://www.mediawiki.org/wiki/Manual:Interwiki))

In that case, we would be back to my first answer, but with a macro that would combine at least two elements: the prefix (key) and the suffix (specific for each call). For simplicity, I would also add a third argument: the text that will be highlighted in the hyperlink.

In that case, you would have, in each page's frontmatter, a 'page variable' declaration like this:

```yaml
---
foo: https://foo.bar.baz/
bar: ....
---
```

What I might do is would be declare the macro like this (super-simplified):


```python
    @env.macro
    def interwiki(text:str, prefix:str, suffix:str) -> str:
          """Create a markdown hyperlink, """
          full_url = f"{prefix}/{suffix}" # join the two
          return f"[{text}]({full_url})"
```

And call it in this way:

```markdown

And this calls {{ interwiki("a special page", foo, 'bar/baz.html') }}

```
(I called it 'interwiki' because that's the way I think about it; but you could name it the way you want, of course).

Is my understanding correct? Would that answer your use case?

**Reply by bradumbaugh** [reply]
Thank you for the fast response! This is generally the right idea, yes. 

But, for example, if I have these documents:

**document-a.md**:

```
---
key: doc-a
url: ...
---

Some text
```

How could I create a link to **document-a.md**, by its `key`, from another document?

**document-b.md**:

```
---
key: doc-b
url: ...
---

{{ link_to('My Link Text', 'doc-a') }}
```

Since the implementation of `link_to`, when called from **document-b.md**, doesn't have access to the front-matter in **document-a.md**, it seems like I can't do this, unless I have extras to use for looking up a URL by `key`.

I'm thinking I need to do something like this in **mkdocs.yml**:

```
extra:
  doc-keys:
    doc-a:
       url: path/to/document-a.md
    doc-b:
       url: path/to/document-b.md
```

In this way, the macro can access the URLs for all documents, by `key`.

I think this might work, I'm just trying to find a way to do it with out the extras, and without having to manually parse the front-matter of all docs each time I invoke the macro.

Please don't feel any pressure to continue with this conversation, you've been extraordinarily helpful and I hate to take your time.

**Reply by fralau** [reply]
Ah OK 🤔  Just curious, why would you adopt this form:

```
---
key: doc-a
url: path/to/document-a.md
---
```

Instead of:

```
---
doc-a: path/to/document-a.md
---
```

But even if it is wordy, I don't see your choice (key/url) as problem!

In any case, if you need to share those keys between documents, you are left with two solutions:

1. Gather them manually in the `extra` section of your config file.
2. Write a Python function, as explained, to read all the frontmatters and collect them in a variable, which will then be accessible to the macros in the markdown pages.

I don't find anything shocking in doing 2. 🤔 

First of all, having to re-scan source code to extract data (but differently) is something that is often necessary and happens quite often time in compilers and interpreters (see, [two-pass](https://www.geeksforgeeks.org/single-pass-two-pass-and-multi-pass-compilers/)). 

Secondly, it should take a very short time, unless your website has hundreds of pages; especially if your read the markdown pages from the `env` object directly (in-memory), instead of the file system.

But perhaps you may want to start with 1, since it is simpler? Then, once you feel more comfortable, you could venture into 2? How you think it is best. 🤷‍♂️

**Reply by bradumbaugh** [reply]
Thanks @fralau! This all makes sense to me, and I think you've got me going in a good direction, here. I'll probably try #2! Thank you very much for the interesting discussion, good ideas, and great library that adds so much value to MkDocs.

**Reply by fralau** [reply]
You're very welcome.
